### DISCUSSION TITLE
Multiple srsUEs latency

### DISCUSSION THREAD

**Discussion by improperaffo** [srsran/srsRAN_Project#546] [General Help] [answered]
Dear all,
we are conducting some tests connecting multiple srsUEs to a single SRS gNB over zmq using the  [GNU Radio script provided in the documentation](https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#multi-ue-emulation).
The connection is reliable and the throughput is as expected. Yet, we are experiencing large latencies:
```
--- 10.45.0.1 ping statistics ---
500 packets transmitted, 500 received, 0% packet loss, time 3876ms
rtt min/avg/max/mdev = 46.902/69.499/132.154/14.478 ms, pipe 17
```

When using only one srsUE and connecting it directly to the gNB (without passing over the GNU Radio script), we can achieve much lower latencies:
```
--- 10.45.0.1 ping statistics ---
500 packets transmitted, 500 packets received, 0% packet loss
round-trip min/avg/max/stddev = 11.302/15.500/26.558/2.794 ms
```

We have tried generating both the python script and the c++ code from GNU Radio, but they both perform in the same way.

Is this the default behavior, or is this a problem of our setup? And how could this problem be solved?

Thank you,
Raphael

P.S. Is anyone working on the support of TDD for srsUE?

**Comment by pgawlowicz** [comment, marked as answer]
Hi, 
there is a throttling element in the gnuradio flowgraph, that slows down the transfer of signal samples. You can speed it up by changing the slow-down factor, if you set 1.0 the samples should be passed with sample rate speed. However, note that the CPU load might be a limiting factor in the case of multiple UEs (there are multiple instances of srsUE process and signal samples have to be delivered and processed by each).
If you connect gnb and srsUE directly, the samples are passes as fast as possible, probably even faster than the sample rate.

**Reply by improperaffo** [reply]
Thank you Piotr, that was the issue. Running it on a powerful CPU, we can achieve these values, which are the ones expected for our setup.
```
--- 10.45.0.1 ping statistics ---
500 packets transmitted, 500 received, 0% packet loss, time 3033ms
rtt min/avg/max/mdev = 10.327/17.180/40.047/4.346 ms, pipe 5
```
